src/operator/batch_norm-inl.h

```diff
@@ -74,154 +103,96 @@ class BatchNormOp : public Operator {
-      // get requested temp space
-      Tensor<xpu, 2> workspace = ctx.requested[batchnorm::kTempSpace].get_space<xpu>(
-          mshadow::Shape2(3, mean.shape_[0]), s);
-      Tensor<xpu, 1> gmean = workspace[0];
-      Tensor<xpu, 1> gvar = workspace[1];
-      Tensor<xpu, 1> tmp = workspace[2];
-
-      moving_mean = moving_mean * param_.momentum + mean * (1 - param_.momentum);
-      moving_var = moving_var * param_.momentum + var * (1 - param_.momentum);
-      // cal
-      gvar = sumall_except_dim<1>((grad * broadcast<1>(slope, data.shape_)) *
-                                  (data - broadcast<1>(mean, data.shape_)) *
-                                  -0.5f *
-                                  F<mshadow_op::power>(broadcast<1>(var + param_.eps, data.shape_),
-                                                       -1.5f));
-      gmean = sumall_except_dim<1>(grad * broadcast<1>(slope, data.shape_));
-      gmean *= -1.0f / F<mshadow_op::square_root>(var + param_.eps);
-      tmp = scale * sumall_except_dim<1>(-2.0f * (data - broadcast<1>(mean, data.shape_)));
-      tmp *= gvar;
-      gmean += tmp;
-      // assign
-      if (!param_.fix_gamma) {
-        Assign(gslope, req[batchnorm::kGamma],
-               sumall_except_dim<1>(
-                   grad * (data - broadcast<1>(mean, data.shape_)) /
-                   F<mshadow_op::square_root>(broadcast<1>(var + param_.eps, data.shape_))));
-      } else {
-        Assign(gslope, req[batchnorm::kGamma], 0.0f);
-      }
-      Assign(grad_in, req[batchnorm::kData],
-             (grad * broadcast<1>(slope, data.shape_)) *
-             broadcast<1>(1.0f / F<mshadow_op::square_root>(var + param_.eps), data.shape_) +
-             broadcast<1>(gvar, data.shape_) * scale * 2.0f * (data - broadcast<1>(mean,
-                                                                                   data.shape_)) +
-             broadcast<1>(gmean, data.shape_) * scale);
-      Assign(gbias, req[batchnorm::kBeta], sumall_except_dim<1>(grad));
-    } else {
-      // use global statistics with freeze moving mean and var.
-      if (!param_.fix_gamma) {
-        Assign(gslope, req[batchnorm::kGamma],
-               sumall_except_dim<1>(
-                   grad * (data - broadcast<1>(moving_mean, data.shape_)) /
-                   F<mshadow_op::square_root>(broadcast<1>(moving_var + param_.eps, data.shape_))));
-      } else {
-        Assign(gslope, req[batchnorm::kGamma], 0.0f);
-      }
-      Assign(gbias, req[batchnorm::kBeta], sumall_except_dim<1>(grad));
-      Assign(grad_in, req[batchnorm::kData], (grad * broadcast<1>(slope, data.shape_)) *
-             broadcast<1>(
-                 1.0f / F<mshadow_op::square_root>(moving_var + param_.eps), data.shape_));
-    }
+    mshadow::Stream<xpu> *s = ctx.get_stream<xpu>();
+    DoBackward(s, ctx, out_grad, in_data,
+               out_data, req, in_grad, aux_states);
   }
 
  private:
+  void DoForward(mshadow::Stream<cpu> *stream,
```

Comment:
As far as I can find, it is not possible to partially specialize the template parameters out-of-line this way (i.e. only specify xpu and not DType and AccReal in the specialization).

Note: 

http://stackoverflow.com/questions/3768862/c-single-template-specialisation-with-multiple-template-parameters

Which would be:  

http://stackoverflow.com/questions/3768862/c-single-template-specialisation-with-multiple-template-parameters

For example:

template <typename F, typename I>
class B
{
 public:
  int x_;
  void someFunc()
  {
    std::cout << "normal" << std::endl;
  };

  void someFuncNotSpecial()
  {
    std::cout << "normal" << std::endl;
  };
};

template <typename I>
class B<float, I>
{
 public:
  void someFunc();
};

template <typename I>
void B<float, I>::someFunc()
{
  x_ = 0;
}

This will compile with x_ not found.